src/org/traccar/protocol/OpenGtsProtocolDecoder.java

```diff
@@ -0,0 +1,141 @@
+            params = decoder.getParameters();
+        }
+
+        Position position = new Position();
+        position.setProtocol(getProtocolName());
+        position.setValid(true);
+
+        for (Map.Entry<String, List<String>> entry : params.entrySet()) {
+            String value = entry.getValue().get(0);
+            switch (entry.getKey()) {
+                case "id":
+                    DeviceSession deviceSession = getDeviceSession(channel, remoteAddress, value);
+                    if (deviceSession == null) {
+                        sendResponse(channel, HttpResponseStatus.BAD_REQUEST);
+                        return null;
+                    }
+                    position.setDeviceId(deviceSession.getDeviceId());
+                    break;
+                case "gprmc":
+                    Parser parser = new Parser(PATTERN, value);
+                    if (!parser.matches()) {
+                            sendResponse(channel, HttpResponseStatus.BAD_REQUEST);
+                        return null;
+                    }
+
+                    DateBuilder dateBuilder = new DateBuilder()
+                            .setTime(parser.nextInt(0), parser.nextInt(0), parser.nextInt(0));
+
+                    position.setValid(parser.next().equals("A"));
+                    position.setLatitude(parser.nextCoordinate());
+                    position.setLongitude(parser.nextCoordinate());
+                    position.setSpeed(parser.nextDouble(0));
+                    position.setCourse(parser.nextDouble(0));
+
+                    dateBuilder.setDateReverse(parser.nextInt(0), parser.nextInt(0), parser.nextInt(0));
+                    position.setTime(dateBuilder.getDate());
+                    break;
+                case "alt":
+                    position.setAltitude(Double.parseDouble(value));
+                    break;
+                case "batt":
+                    position.set(Position.KEY_BATTERY_LEVEL, Double.parseDouble(value));
+                    break;
+                default:
+                    try {
+                        position.set(entry.getKey(), Double.parseDouble(value));
+                    } catch (NumberFormatException e) {
+                        switch (value) {
+                            case "true":
+                                position.set(entry.getKey(), true);
+                                break;
+                            case "false":
+                                position.set(entry.getKey(), false);
+                                break;
+                            default:
+                                position.set(entry.getKey(), value);
+                                break;
+                        }
+                    }
+                    break;
+            }
+        }
+
+        if (position.getFixTime() == null) {
+            position.setTime(new Date());
```

Comment:
Why do we need this?

Reply:
Yes, looks like this is irrelevant. Removed.